File: Chapter 14/JS/section_14_6_instrumentation_logging/tests/orderTelemetryQualityGateTests.js
/**
 * Listing 14.5 — Automated Telemetry Quality Gate
 *
 * Book file: com/ecommerce/order/tests/OrderTelemetryQualityGateTests.java (inlined)
 */
import { test } from "node:test";
import assert from "node:assert";
import { PaymentPort } from "../ports/paymentPort.js";
import { OrderService, orderIdContext } from "../services/orderService.js";

test("checkout should maintain MDC context boundary during execution", () => {
  // Arrange: Intercept interface execution to read thread-local variables
  let capturedOrderId = null;

  const customInterceptorPort = new PaymentPort();
  customInterceptorPort.process = (amount) => {
    // Read active thread context values mid-transaction
    capturedOrderId = orderIdContext.getStore()?.orderId ?? null;
    return true;
  };

  const service = new OrderService(customInterceptorPort);

  // Act: Trigger the system transaction path
  service.checkout("ord_99812", 75.0);

  // Assert: Context was present mid-transaction
  assert.strictEqual(
    capturedOrderId,
    "ord_99812",
    "Telemetry Gap Error: MDC Context was dropped before crossing the port boundary!"
  );

  // Assert: Ensure clean context teardown to prevent memory leaks
  assert.strictEqual(
    orderIdContext.getStore() ?? null,
    null,
    "Memory Contamination Error: MDC context leaked past the request boundary lifetime!"
  );
});

File: Chapter 14/Java/src/test/java/com/grokkingsoftwarearchitecture/chapter14/section_14_6_instrumentation_logging/OrderTelemetryQualityGateTests.java
package com.grokkingsoftwarearchitecture.chapter14.section_14_6_instrumentation_logging;

import com.grokkingsoftwarearchitecture.chapter14.section_14_6_instrumentation_logging.ports.PaymentPort;
import com.grokkingsoftwarearchitecture.chapter14.section_14_6_instrumentation_logging.services.OrderService;
import org.junit.jupiter.api.Test;
import org.slf4j.MDC;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

/**
 * Listing 14.5 — Automated Telemetry Quality Gate
 *
 * Book file: com/ecommerce/order/tests/OrderTelemetryQualityGateTests.java (inlined)
 */
public class OrderTelemetryQualityGateTests {

    @Test
    void testCheckout_Should_MaintainMdcContextBoundaryDuringExecution() {
        // Arrange: Intercept interface execution to read Thread-Local variables
        PaymentPort customInterceptorPort = amount -> {
            // Read active thread context values mid-transaction
            String activeOrderId = MDC.get("orderId");
            assertEquals("ord_99812", activeOrderId,
                "Telemetry Gap Error: MDC Context was dropped before crossing the port boundary!");
            return true;
        };

        OrderService service = new OrderService(customInterceptorPort);

        // Act: Trigger the system transaction path
        service.checkout("ord_99812", 75.00);

        // Assert: Ensure clean thread teardown to prevent memory context leaks
        assertNull(MDC.get("orderId"),
            "Memory Contamination Error: MDC context leaked past the request boundary lifetime!");
    }
}